### PR TITLE
Saving the muon analysis 2 GUI

### LINKED ISSUE #19150: Saving the State of Muon Analysis
The current version of Mantid (3.9.1) cannot save the state of the muon analysis interface (along with some other custom interfaces). Being able to save the state of the interface would present users with a more pleasant experience. 

To achieve this the Muon Analysis code will be adapted to allow for the saving of the current state of the interface. 

It has been advised to discuss this with Sam.

### PR DESCRIPTION
**Pull request by AnthonyLim23**
This PR is to add the ability to save a python GUI (muon analysis 2) as part of project saving. I will add some documentation on how to save a GUI in #23836 and it does not work with project recovery #23837

**To test:**
Before compiling you will need to enable muon analysis 2 by adding:
`Muon/Muon_Analysis_2.py` to https://github.com/mantidproject/mantid/blob/master/Framework/Properties/Mantid.properties.template#L23

Once it has built:
1. Open Mantid and load any workspace
2. Open Muon Analysis 2
3. Edit the interface state (change group names etc.)
4. Save the project and make sure to include both the ws and GUI
5. Close Mantid
6. Open Mantid and load in the project we just saved
7. Muon Analysis 2 should open with the values you had previously entered. 

If you feeling adventures you can go into the project file and change some of the values. 

Fixes #19150. 

---
This does not require release notes
#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AnthonyLim23** [comment]
it seems to have stopped working. I am currently investigating and checking it across platforms.

**Comment by AnthonyLim23** [comment]
this is ready for review now

File: scripts/Muon_Analysis_2.py [hunk 1]
```diff
@@ -66,31 +79,44 @@ def qapp():
...
-        QtGui.QMessageBox.warning(muonGUI, "Muon Analysis version 2", str(error))
-        return muonGUI
+        muon = QtGui.QWidget()
+        QtGui.QMessageBox.warning(muon, Name, str(error))
+        return muon
 
 
 def saveToProject():
-    if muonGUI is None:
+    print("waaa")
```
Comment (martyngigg) [inline scripts/Muon_Analysis_2.py]: I assume this line was left in by accident?

File: scripts/pythonTSV.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    if isinstance(value, int):
+        TSV.storeInt(value)
+    elif isinstance(value, float):
+        TSV.storeDouble(value)
+    elif isinstance(value, bool):
+        TSV.storeBool(value)
+    elif isinstance(value, str):
+        TSV.storeString(value)
+    else:
+        raise TypeError
```
Comment (martyngigg) [inline scripts/pythonTSV.py]: Can we raise the exception with a message giving some idea of the issue?

File: scripts/pythonTSV.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    if isinstance(value, int):
+        return TSV.readInt()
+    elif isinstance(value, float):
+        return TSV.readDouble()
+    elif isinstance(value, bool):
+        return TSV.readBool()
+    elif isinstance(value, str):
+        return TSV.readString()
+    else:
+        raise TypeError
```
Comment (martyngigg) [inline scripts/pythonTSV.py]: Can we raise the exception with a message giving some idea of the issue?

File: scripts/Muon/GUI/Common/muon_context/muon_context.py [hunk 4]
```diff
@@ -49,9 +56,84 @@ def printContext(self):
         print(LoadText, self.common_context[LoadText])
         print(Groups)
         for group in self.common_context[Groups]:
-            print(group.name)
+            group.Print()
```
Comment (martyngigg) [inline scripts/Muon/GUI/Common/muon_context/muon_context.py]: Is there a reason we need to print everything as we save?

**Review by martyngigg — COMMENTED** [review]
The saving and loading seem to work as expected.

I've made a couple of inline comments

File: scripts/Muon/GUI/Common/muon_context/muon_context.py (hunk 4 above)
Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/muon_context/muon_context.py]: this is not for saving and is just part of the print context method.

Comment (martyngigg) [reply scripts/Muon/GUI/Common/muon_context/muon_context.py]: The print statements seem to come out in the results log when it saves though.

Comment (AnthonyLim23) [reply scripts/Muon/GUI/Common/muon_context/muon_context.py]: should be fixed now, I had a print statement inside of the save method

**Comment by AnthonyLim23** [comment]
looks like there is a build error with conflicting qt versions in the test. 

```
17:59:05     from mantidqtpython import MantidQt
17:59:05 RuntimeError: the PyQt4.QtCore and PyQt5.QtCore modules both wrap the QObject class
```

**Review by martyngigg — APPROVED** [review]
This looks good now the macOS error is fixed.

**Pull request merged**